File: dev/NetUtil.js
"use strict"

class NetUtil {

    static format (value, type="string") {
        switch (true) {

            case type=="string" && typeof value=="string":
                value = value.replace(/(_|\s)/g, "").toLowerCase()
                break

            case type=="time" && typeof value=="number":
                const date = new Date(value)
                const formatted = []

                if (value < 1000) {
                    formatted.push(`${date.getMilliseconds()}ms`)

                } else {

                    if (value >= 3600000) formatted.push(`${date.getHours()}h`)
                    if (value >= 60000)   formatted.push(`${date.getMinutes()}m`)

                    formatted.push(`${date.getSeconds()}s`)
                }

                value = formatted.join(" ")
                break
        }

        return value
    }

    static shuffle (arr) {
        for (let i=arr.length; i; i--) {
            const j = Math.floor(Math.random() * i)
            const x = arr[i-1]
            arr[i-1] = arr[j]
            arr[j] = x
        }
    }

    static addZeroPadding (map, zP) {
        const extraColumns = [...new Array(zP)].map(v => 0)
        map = map.map(row => [...extraColumns, ...row, ...extraColumns])

        const extraRows = [...new Array(zP)].map(r => [...new Array(map.length+zP*2)].map(x => 0))
        return [...extraRows.slice(0), ...map, ...extraRows.slice(0)]
    }

    static arrayToMap (arr, size) {
        const map = []

        for (let i=0; i<size; i++) {
            map[i] = []

            for (let j=0; j<size; j++) {
                map[i][j] = arr[i*size+j]
            }
        }

        return map
    }

    static arrayToVolume (arr, channels) {

        const vol = []
        const size = Math.sqrt(arr.length/channels)
        const mapValues = size**2

        for (let d=0; d<Math.floor(arr.length/mapValues); d++) {

            const map = []

            for (let i=0; i<size; i++) {
                map[i] = []

                for (let j=0; j<size; j++) {
                    map[i][j] = arr[d*mapValues  + i*size+j]
                }
            }

            vol[d] = map
        }

        return vol
    }

    static convolve ({input, zeroPadding, weights, channels, stride, bias}) {

        const inputVol = NetUtil.arrayToVolume(input, channels)
        const outputMap = []

        const paddedLength = inputVol[0].length + zeroPadding*2
        const fSSpread = Math.floor(weights[0].length / 2)

        // For each input channels,
        for (let di=0; di<channels; di++) {
            inputVol[di] = NetUtil.addZeroPadding(inputVol[di], zeroPadding)
            // For each inputY without ZP
            for (let inputY=fSSpread; inputY<paddedLength-fSSpread; inputY+=stride) {
                outputMap[(inputY-fSSpread)/stride] = outputMap[(inputY-fSSpread)/stride] || []
                // For each inputX without zP
                for (let inputX=fSSpread; inputX<paddedLength-fSSpread; inputX+=stride) {
                    let sum = 0
                    // For each weightsY on input
                    for (let weightsY=0; weightsY<weights[0].length; weightsY++) {
                        // For each weightsX on input
                        for (let weightsX=0; weightsX<weights[0].length; weightsX++) {
                            sum += inputVol[di][inputY+(weightsY-fSSpread)][inputX+(weightsX-fSSpread)] * weights[di][weightsY][weightsX]
                        }
                    }

                    outputMap[(inputY-fSSpread)/stride][(inputX-fSSpread)/stride] = (outputMap[(inputY-fSSpread)/stride][(inputX-fSSpread)/stride]||0) + sum
                }
            }
        }

        // Then add bias
        for (let outY=0; outY<outputMap.length; outY++) {
            for (let outX=0; outX<outputMap.length; outX++) {
                outputMap[outY][outX] += bias
            }
        }

        return outputMap
    }

    static buildConvErrorMap (layer, filter, filterI) {

        // Clear the existing error values, first
        for (let row=0; row<filter.errorMap.length; row++) {
            for (let col=0; col<filter.errorMap[0].length; col++) {
                filter.errorMap[row][col] = 0
            }
        }

        // Cache / convenience
        const zeroPadding = layer.nextLayer.zeroPadding

        const fSSpread = Math.floor(layer.nextLayer.filterSize / 2)
        const paddedLength = filter.errorMap.length + zeroPadding*2

        // Zero pad the error map, to allow easy convoling
        // TODO, may be more performant to just use if statements when updating, instead
        filter.errorMap = NetUtil.addZeroPadding(filter.errorMap, zeroPadding)

        // For each channel in filter in the next layer which corresponds to this filter
        for (let nlFilterI=0; nlFilterI<layer.nextLayer.size; nlFilterI++) {

            const weights = layer.nextLayer.filters[nlFilterI].weights[filterI]
            const errorMap = layer.nextLayer.filters[nlFilterI].errorMap

            // Unconvolve their error map using the weights
            for (let inputY=fSSpread; inputY<paddedLength - fSSpread; inputY+=layer.nextLayer.stride) {
                for (let inputX=fSSpread; inputX<paddedLength - fSSpread; inputX+=layer.nextLayer.stride) {

                    for (let weightsY=0; weightsY<layer.nextLayer.filterSize; weightsY++) {
                        for (let weightsX=0; weightsX<layer.nextLayer.filterSize; weightsX++) {

                            filter.errorMap[inputY+(weightsY-fSSpread)][inputX+(weightsX-fSSpread)] += weights[weightsY][weightsX]
                                * errorMap[(inputY-fSSpread)/layer.nextLayer.stride][(inputX-fSSpread)/layer.nextLayer.stride]
                        }
                    }
                }
            }
        }

        // Take out the zero padding. Rows:
        filter.errorMap = filter.errorMap.splice(zeroPadding, filter.errorMap.length - zeroPadding*2)

        // Columns:
        for (let emXI=0; emXI<filter.errorMap.length; emXI++) {
            filter.errorMap[emXI] = filter.errorMap[emXI].splice(zeroPadding, filter.errorMap[emXI].length - zeroPadding*2)
        }
    }

    static buildConvDWeights (layer) {

        const weightsCount = layer.filters[0].weights[0].length
        const fSSpread = Math.floor(weightsCount / 2)
        const channelsCount = layer.filters[0].weights.length

        // Adding an intermediary step to allow regularization to work
        const deltaDeltaWeights = []

        // Filling the deltaDeltaWeights with 0 values
        for (let weightsY=0; weightsY<weightsCount; weightsY++) {
            deltaDeltaWeights[weightsY] = []
            for (let weightsX=0; weightsX<weightsCount; weightsX++) {
                deltaDeltaWeights[weightsY][weightsX] = 0
            }
        }

        // For each filter
        for (let filterI=0; filterI<layer.filters.length; filterI++) {

            const filter = layer.filters[filterI]

            // Each channel will take the error map and the corresponding inputMap from the input...
            for (let channelI=0; channelI<channelsCount; channelI++) {

                const inputValues = NetUtil.getActivations(layer.prevLayer, channelI, layer.inMapValuesCount)
                const inputMap = NetUtil.addZeroPadding(NetUtil.arrayToMap(inputValues, Math.sqrt(layer.inMapValuesCount)), layer.zeroPadding)

                // ...slide the filter with correct stride across the zero-padded inputMap...
                for (let inputY=fSSpread; inputY<inputMap.length-fSSpread; inputY+=layer.stride) {
                    for (let inputX=fSSpread; inputX<inputMap.length-fSSpread; inputX+=layer.stride) {

                        // ...and at each location...
                        for (let weightsY=0; weightsY<weightsCount; weightsY++) {
                            for (let weightsX=0; weightsX<weightsCount; weightsX++) {

                                const activation = inputMap[inputY-fSSpread+weightsY][inputX-fSSpread+weightsX]

                                // Increment and regularize the delta delta weights by the input activation (later multiplied by the error)
                                deltaDeltaWeights[weightsY][weightsX] += activation *
                                     (1 + (((layer.net.l2||0)+(layer.net.l1||0))/layer.net.miniBatchSize) * filter.weights[channelI][weightsY][weightsX])
                            }
                        }

                        const error = filter.errorMap[(inputY-fSSpread)/layer.stride][(inputX-fSSpread)/layer.stride]

                        // Applying and resetting the deltaDeltaWeights
                        for (let weightsY=0; weightsY<weightsCount; weightsY++) {
                            for (let weightsX=0; weightsX<weightsCount; weightsX++) {
                                filter.deltaWeights[channelI][weightsY][weightsX] += deltaDeltaWeights[weightsY][weightsX] * error
                                deltaDeltaWeights[weightsY][weightsX] = 0
                            }
                        }
                    }
                }
            }

            // Increment the deltaBias by the sum of all errors in the filter
            for (let eY=0; eY<filter.errorMap.length; eY++) {
                for (let eX=0; eX<filter.errorMap.length; eX++) {
                    filter.deltaBias += filter.errorMap[eY][eX]
                }
            }
        }
    }


    static getActivations (layer, mapStartI, mapSize){

        if (arguments.length==1) {

            const returnArr = []

            if (layer instanceof FCLayer) {

                for (let ni=0; ni<layer.neurons.length; ni++) {
                    returnArr.push(layer.neurons[ni].activation)
                }
            } else {

                for (let fi=0; fi<layer.filters.length; fi++) {
                    for (let rowI=0; rowI<layer.filters[fi].activationMap.length; rowI++) {
                        for (let colI=0; colI<layer.filters[fi].activationMap[rowI].length; colI++) {
                            returnArr.push(layer.filters[fi].activationMap[rowI][colI])
                        }
                    }
                }
            }

            return returnArr
        } else {

            const returnArr = []

            if (layer instanceof FCLayer) {

                for (let i=mapStartI*mapSize; i<(mapStartI+1)*mapSize; i++) {
                    returnArr.push(layer.neurons[i].activation)
                }
            } else {

                for (let row=0; row<layer.filters[mapStartI].activationMap.length; row++) {
                    for (let col=0; col<layer.filters[mapStartI].activationMap[row].length; col++) {
                        returnArr.push(layer.filters[mapStartI].activationMap[row][col])
                    }
                }
            }
            return returnArr
        }
    }
}

typeof window=="undefined" && (exports.NetUtil = NetUtil)

File: dev/cpp/NetUtil.cpp

void NetUtil::shuffle (std::vector<std::tuple<std::vector<double>, std::vector<double> > > &values) {
    for (int i=values.size(); i; i--) {
        int j = floor(rand() / RAND_MAX * i);
        std::tuple<std::vector<double>, std::vector<double> > x = values[i-1];
        values[i-1] = values[j];
        values[j] = x;
    }
}

std::vector<std::vector<double> > NetUtil::addZeroPadding (std::vector<std::vector<double> > map, int zP) {

    // Left and right columns
    for (int row=0; row<map.size(); row++) {
        for (int z=0; z<zP; z++) {
            map[row].insert(map[row].begin(), 0.0);
            map[row].push_back(0);
        }
    }

    // Top rows
    for (int z=0; z<zP; z++) {
        std::vector<double> row;

        for (int i=0; i<map[0].size(); i++) {
            row.push_back(0);
        }

        map.insert(map.begin(), row);
    }

    // Bottom rows
    for (int z=0; z<zP; z++) {
        std::vector<double> row;

        for (int i=0; i<map[0].size(); i++) {
            row.push_back(0);
        }

        map.push_back(row);
    }

    return map;
}

std::vector<std::vector<double> > NetUtil::arrayToMap (std::vector<double> array, int size) {

    std::vector<std::vector<double> > map;

    for (int i=0; i<size; i++) {
        std::vector<double> row;

        for (int j=0; j<size; j++) {
            row.push_back(array[i*size+j]);
        }

        map.push_back(row);
    }

    return map;
}

std::vector<std::vector<std::vector<double> > > NetUtil::arrayToVolume (std::vector<double> array, int channels) {

    std::vector<std::vector<std::vector<double> > > vol;
    int size = sqrt(array.size() / channels);
    int mapValues = size * size;

    for (int d=0; d<floor(array.size() / mapValues); d++) {
        std::vector<std::vector<double> > map;

        for (int i=0; i<size; i++) {
            std::vector<double> row;

            for (int j=0; j<size; j++) {
                row.push_back(array[d*mapValues + i*size+j]);
            }
            map.push_back(row);
        }

        vol.push_back(map);
    }

    return vol;
}

std::vector<std::vector<double> > NetUtil::convolve(std::vector<std::vector<std::vector<double> > > input, int zP,
    std::vector<std::vector<std::vector<double> > > weights, int channels, int stride, double bias) {

    std::vector<std::vector<double> > output;

    int outSize = (input[0].size() - weights[0].size() + 2*zP) / stride + 1;

    // Fill with 0 values
    for (int r=0; r<outSize; r++) {
        std::vector<double> row;
        for (int v=0; v<outSize; v++) {
            row.push_back(0.0);
        }
        output.push_back(row);
    }

    int x = -zP;
    int y = -zP;

    for (int outY=0; outY<outSize; y+=stride, outY++) {

        x = -zP;

        for (int outX=0; outX<outSize; x+=stride, outX++) {

            double sum = 0;

            for (int weightsY=0; weightsY<weights[0].size(); weightsY++) {

                int inputY = y+weightsY;

                for (int weightsX=0; weightsX<weights[0].size(); weightsX++) {

                    int inputX = x+weightsX;

                    if (inputY>=0 && inputY<input[0].size() && inputX>=0 && inputX<input[0].size()) {
                        for (int di=0; di<channels; di++) {
                            sum += input[di][inputY][inputX] * weights[di][weightsY][weightsX];
                        }
                    }
                }
            }

            sum += bias;

            output[outY][outX] = sum;
        }
    }

    return output;
}

template <class T>
std::vector<std::vector<std::vector<T> > > NetUtil::createVolume (int depth, int rows, int columns, T value) {

    std::vector<std::vector<std::vector<T> > > volume;

    for (int d=0; d<depth; d++) {
        std::vector<std::vector<T> > map;

        for (int r=0; r<rows; r++) {
            std::vector<T> row;

            for (int c=0; c<columns; c++) {
                row.push_back(value);
            }

            map.push_back(row);
        }

        volume.push_back(map);
    }

    return volume;
}

std::vector<std::vector<double> > NetUtil::buildConvErrorMap (int paddedLength, Layer* nextLayer, int filterI) {

    // Cache / convenience
    int zeroPadding = nextLayer->zeroPadding;
    int fsSpread = floor(nextLayer->filterSize / 2);

    std::vector<std::vector<double> > errorMap;

    // Zero pad and clear the error map, to allow easy convoling
    for (int row=0; row<paddedLength; row++) {
        std::vector<double> paddedRow;

        for (int val=0; val<paddedLength; val++) {
            paddedRow.push_back(0);
        }

        errorMap.push_back(paddedRow);
    }

    // For each channel in filter in the next layer which corresponds to this filter
    for (int nlFilterI=0; nlFilterI<nextLayer->filters.size(); nlFilterI++) {

        std::vector<std::vector<double> > weights = nextLayer->filterWeights[nlFilterI][filterI];
        std::vector<std::vector<double> > errMap = nextLayer->errors[nlFilterI];

        // Unconvolve their error map using the weights
        for (int inY=fsSpread; inY<paddedLength - fsSpread; inY+=nextLayer->stride) {
            for (int inX=fsSpread; inX<paddedLength - fsSpread; inX+=nextLayer->stride) {

                for (int wY=0; wY<nextLayer->filterSize; wY++) {
                    for (int wX=0; wX<nextLayer->filterSize; wX++) {
                        errorMap[inY+(wY-fsSpread)][inX+(wX-fsSpread)] += weights[wY][wX]
                            * errMap[(inY-fsSpread)/nextLayer->stride][(inX-fsSpread)/nextLayer->stride];
                    }
                }
            }
        }
    }

    // Take out the zero padding. Rows:
    errorMap.erase(errorMap.begin(), errorMap.begin()+zeroPadding);
    errorMap.erase(errorMap.end()-zeroPadding, errorMap.end());

    // Columns:
    for (int eY=0; eY<errorMap.size(); eY++) {
        errorMap[eY].erase(errorMap[eY].begin(), errorMap[eY].begin()+zeroPadding);
        errorMap[eY].erase(errorMap[eY].end()-zeroPadding, errorMap[eY].end());
    }

    return errorMap;
}

void NetUtil::buildConvDWeights (ConvLayer* layer) {

    int weightsCount = layer->filterWeights[0][0].size();
    int fsSpread = floor(weightsCount / 2);
    int channelsCount = layer->filterWeights[0].size();

    // For each filter
    for (int f=0; f<layer->filters.size(); f++) {

        // Each channel will take the error map and the corresponding inputMap from the input...
        for (int c=0; c<channelsCount; c++) {

            std::vector<double> inputValues = NetUtil::getActivations(layer->prevLayer, c, layer->inMapValuesCount);
            std::vector<std::vector<double> > inputMap = NetUtil::addZeroPadding(NetUtil::arrayToMap(inputValues, sqrt(layer->inMapValuesCount)), layer->zeroPadding);

            // ...slide the filter with correct stride across the zero-padded inputMap...
            for (int inY=fsSpread; inY<inputMap.size()-fsSpread; inY+= layer->stride) {
                for (int inX=fsSpread; inX<inputMap.size()-fsSpread; inX+= layer->stride) {

                    double error = layer->errors[f][(inY-fsSpread)/layer->stride][(inX-fsSpread)/layer->stride];

                    // ...and at each location...
                    for (int wY=0; wY<weightsCount; wY++) {
                        for (int wX=0; wX<weightsCount; wX++) {
                            // activation * error
                            layer->filterDeltaWeights[f][c][wY][wX] += inputMap[inY-fsSpread+wY][inX-fsSpread+wX] * error;
                        }
                    }
                }
            }
        }

        // Increment the deltaBias by the sum of all errors in the filter
        for (int eY=0; eY<layer->errors[f].size(); eY++) {
            for (int eX=0; eX<layer->errors[f].size(); eX++) {
                layer->deltaBiases[f] += layer->errors[f][eY][eX];
            }
        }
    }
}

std::vector<double> NetUtil::getActivations (Layer* layer, int mapStartI, int mapSize) {

    std::vector<double> activations;

    if (layer->type == "FC") {

        for (int n=mapStartI*mapSize; n<(mapStartI+1)*mapSize; n++) {
            activations.push_back(layer->actvns[n]);
        }

    } else if (layer->type == "Conv") {

        for (int r=0; r<layer->activations[mapStartI].size(); r++) {
            for (int c=0; c<layer->activations[mapStartI][r].size(); c++) {
                activations.push_back(layer->activations[mapStartI][r][c]);
            }
        }

    } else {

        for (int r=0; r<layer->activations[mapStartI].size(); r++) {
            for (int v=0; v<layer->activations[mapStartI].size(); v++) {
                activations.push_back(layer->activations[mapStartI][r][v]);
            }
        }

    }

    return activations;
}